astropy/table/table.py

```diff
@@ -763,9 +763,12 @@ def _set_masked_from_cols(self, cols):
                 self._set_masked(True)
 
     def _init_from_list_of_dicts(self, data, names, dtype, n_cols, copy):
-        names_from_data = set()
-        for row in data:
-            names_from_data.update(row)
+        if isinstance(data[0], OrderedDict):
```

Comment:
The logic here is not entirely OK because it skips the iteration through every data row to check the set (union) of all keys in the row dicts.  In other words, the first row might have 'a' and 'b' but then other rows have 'a', 'b', 'c'.  This should produce an exception, but the way it is done in this patch that would not happen.  So should be the original lines of code:
```
names_from_data = set()
for row in data:
    names_from_data.update(row)
names_from_data = sorted(names_from_data)
```
Followed by a new check:
```
if isinstance(data[0], OrderedDict) and set(data[0].keys()) == set(names_from_data):
    names_from_data = list(data[0].keys())
# Note: if set(data[0].keys()) != names_from_data this will give an exception later
# so no need to catch here.
```